compiler/rustc_mir/src/transform/generator.rs

```diff
@@ -1340,10 +1340,13 @@ impl<'tcx> MirPass<'tcx> for StateTransform {
         transform.visit_body(body);
 
         // Update our MIR struct to reflect the changes we've made
-        body.yield_ty = None;
         body.arg_count = 2; // self, resume arg
         body.spread_arg = None;
-        body.generator_layout = Some(layout);
+
+        body.generator.as_mut().map(|gen| {
+            gen.yield_ty = None;
+            gen.generator_layout = layout;
+        });
```

Comment:
You can use `.unwrap()` here and below as this pass immediately returns for non-generator types.

Reply:
I wanted to avoid multiple unwraps and a mutable borrow would upset the borrow checker. I can change it, but doesn't matter until we know the perf results.